### PR TITLE
JP-3649 Fix cube build error when called from mrs_imatch and add deprecated warning for mrs_imatch

### LINKED ISSUE #8538: Spec3 step mrs_imatch crashing
_Issue [JP-3649](https://jira.stsci.edu/browse/JP-3649) was created on JIRA by [Maria Pena-Guerrero](https://jira.stsci.edu/secure/ViewProfile.jspa?name=pena):_

The mrs_imatch step is skipped by default for data with USEAFTER 2022-05-01. I used the test set that is currently in artifactory for regression tests for MIRI MRS. When it runs it crashes with the following traceback:

Traceback (most recent call last):
  File "/Users/pena/Documents/SCSB/analysis_tools/analysis_tools/src/run_st_pipes.py", line 713, in run_st_pipes
    run_spec3(asn_json, outdir, save_stps_lst=save_stps_lst,
  File "/Users/pena/Documents/SCSB/analysis_tools/analysis_tools/src/run_st_pipes.py", line 418, in run_spec3
    spec3.call(asn_json, output_dir=output_dir, steps=parameter_dict, save_results=True)
  File "/Users/pena/Documents/SCSB/stpipe/stpipe/src/stpipe/step.py", line 686, in call
    return instance.run(*args)
           ^^^^^^^^^^^^^^^^^^^
  File "/Users/pena/Documents/SCSB/stpipe/stpipe/src/stpipe/step.py", line 509, in run
    step_result = self.process(*args)
                  ^^^^^^^^^^^^^^^^^^^
  File "/Users/pena/Documents/SCSB/jwst/jwst/jwst/pipeline/calwebb_spec3.py", line 221, in process
    result = self.mrs_imatch(result)
             ^^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/pena/Documents/SCSB/stpipe/stpipe/src/stpipe/step.py", line 509, in run
    step_result = self.process(*args)
                  ^^^^^^^^^^^^^^^^^^^
  File "/Users/pena/Documents/SCSB/jwst/jwst/jwst/mrs_imatch/mrs_imatch_step.py", line 85, in process
    _match_models(single_ch[c], channel=str(c), degree=degree)
  File "/Users/pena/Documents/SCSB/jwst/jwst/jwst/mrs_imatch/mrs_imatch_step.py", line 266, in _match_models
    cube_models = cbs.process(models)
                  ^^^^^^^^^^^^^^^^^^^
  File "/Users/pena/Documents/SCSB/jwst/jwst/jwst/cube_build/cube_build_step.py", line 369, in process
    cube_container = thiscube.build_ifucube_single()
                     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/pena/Documents/SCSB/jwst/jwst/jwst/cube_build/ifu_cube.py", line 827, in build_ifucube_single
    coord1, coord2, corner_coord, wave, dwave, flux, err, slice_no, \
ValueError: too many values to unpack (expected 13)
None

### PR DESCRIPTION
**Pull request by jemorrison**
Resolves [JP-3649](https://jira.stsci.edu/browse/JP-3649)

Closes #8538 

This PR addresses a bug in cube_build when it was called from  mrs_imatch. Mrs_imatch  is not used in standard processing. A deprecation warning has been added if mrs_imatch is called and the step has been set to default=True in spec. This will prevent the step from being called when the .run method is used for spec3. 

**Checklist for PR authors (skip items if you don't have permissions or they are not applicable)**
- [x] added entry in `CHANGES.rst` within the relevant release section
- [x] updated or added relevant tests
- [x] updated relevant documentation
- [x] added relevant milestone
- [x] added relevant label(s)
- [x] ran regression tests, post a link to the Jenkins job below.
      [How to run regression tests on a PR](https://github.com/spacetelescope/jwst/wiki/Running-Regression-Tests-Against-PR-Branches)
      Running regression tests:
https://plwishmaster.stsci.edu:8081/job/RT/job/JWST-Developers-Pull-Requests/1681/

- [ ] All comments are resolved
- [ ] Make sure the JIRA ticket is [resolved properly](https://github.com/spacetelescope/jwst/wiki/How-to-resolve-JIRA-issues)

### CODE REVIEW

**Comment by jemorrison** [comment]
regression tests: https://plwishmaster.stsci.edu:8081/job/RT/job/JWST-Developers-Pull-Requests/1678/

File: CHANGES.rst [hunk 1]
```diff
@@ -22,6 +22,7 @@ cube_build
 
 - Removed direct setting of the ``self.skip`` attribute from within the step
   itself. [#8600]
+- Fixed a bug when cube_build was called from the mrs_imatch step. [#8728]
```
Comment (melanieclarke) [inline CHANGES.rst:25]:
```suggestion

- Fixed a bug when ``cube_build`` was called from the ``mrs_imatch`` step. [#8728]
```

File: docs/jwst/mrs_imatch/description.rst [hunk 2]
```diff
@@ -8,6 +8,10 @@ Description
 
 Overview
 --------
+
+This step has been deprecated in JWST calibration pipeline version 11.1.
+The mrs_imatch step is skipped in the default running of the calwebb_spec3 pipeline.
```
Comment (melanieclarke) [inline docs/jwst/mrs_imatch/description.rst:13]:
```suggestion
The ``mrs_imatch`` step is skipped by default in the calwebb_spec3 pipeline.
```

File: jwst/mrs_imatch/mrs_imatch_step.py [hunk 3]
```diff
@@ -29,16 +29,18 @@ class MRSIMatchStep(Step):
...
     """
 
     reference_file_types = []
 
     def process(self, images):
-        all_models2d = ModelContainer(images)
 
-        chm = {}
+        # Provide warning to user this code is deprecated
+        warnings.warn("mrs_imatch deprecated", RuntimeWarning)
```
Comment (melanieclarke) [inline jwst/mrs_imatch/mrs_imatch_step.py:40]: I'm not sure this needs to be a software warning - maybe just log a warning instead, for user information?

**Review by melanieclarke — COMMENTED** [review]
Fix looks good to me by eye.  I'll see if I can find some real data to check it on.

The unit test is probably failing because the step is now skipped by default - you'll need to update the test to set skip=False.

Maybe we should also add a quick unit test to verify that the step is skipped by default, and output has status SKIPPED?

**Comment by melanieclarke** [comment]
> I'll see if I can find some real data to check it on.

Just following up - I tested it on the data from the linked helpdesk ticket and it looks good.  The step is skipped by default, completes successfully, with a warning, when manually set to run.

File: CHANGES.rst (hunk 1 above)
Comment (jemorrison) [reply CHANGES.rst:25]: done

File: docs/jwst/mrs_imatch/description.rst (hunk 2 above)
Comment (jemorrison) [reply docs/jwst/mrs_imatch/description.rst:13]: done

File: jwst/mrs_imatch/mrs_imatch_step.py (hunk 3 above)
Comment (jemorrison) [reply jwst/mrs_imatch/mrs_imatch_step.py:40]: changed to log.warning if run

**Comment by jemorrison** [comment]
@melanieclarke I modified the line in the test test_imatch_background_subtracted  so that mrs_imatch would run (well the was my plan).
 with pytest.raises(ValueError):
        step = MRSIMatchStep(skip=False)
        step.call(new_container)
        
       I also tried
  with pytest.raises(ValueError):
        step = MRSIMatchStep()
        step.skip=False
        step.call(new_container)
        
        Neither one of them resulted in running the mrs_imatch step. 
        what am I doing wrong ? 
        I say that because this test should pass if it was run (RIGHT?) 
        But I get a failure  - which I am assuming was because it was not run

**Comment by melanieclarke** [comment]
> what am I doing wrong ?

I think this is probably because you are trying to set the parameter in the step instance.  Calling `step.run()` will respect parameters set that way but calling step.call() throws them away.  Try `MRSIMatchStep.call(new_container, skip=False)` or `step.run()` after setting `step.skip = False`.

**Comment by jemorrison** [comment]
@regression test run have failures but I do not think they are related to this PR.

File: CHANGES.rst [hunk 4]
```diff
@@ -28,6 +28,7 @@ cube_build
 
 - Removed direct setting of the ``self.skip`` attribute from within the step
   itself. [#8600]
+- Fixed a bug when ``cube_build`` was called from the ``mrs_imatch`` step. [#8728]
```
Comment (melanieclarke) [inline CHANGES.rst:31]: This needs a newline above, for formatting consistency.

File: docs/jwst/mrs_imatch/description.rst [hunk 5]
```diff
@@ -8,6 +8,10 @@ Description
 
 Overview
 --------
+
+This step has been deprecated in JWST calibration pipeline version 11.1.
```
Comment (melanieclarke) [inline docs/jwst/mrs_imatch/description.rst:12]: Sorry, I missed this earlier.  The build will be 11.1, the pipeline version is not yet known.  Maybe we should revise this to just say that the step is deprecated and will be skipped by default?  They can refer to the change log if they need to know when it became deprecated.

**Review by melanieclarke — COMMENTED** [review]
Looks good now!  A couple more tiny comments on the change log and docs, and I'll check out the regtest failures.

File: docs/jwst/mrs_imatch/description.rst (hunk 5 above)
Comment (jemorrison) [reply docs/jwst/mrs_imatch/description.rst:12]: done

File: CHANGES.rst (hunk 4 above)
Comment (jemorrison) [reply CHANGES.rst:31]: done

**Review by melanieclarke — APPROVED** [review]
Okay, I think this is good to go now.  I'll merge when CI completes.

**Pull request merged**